Sources/PackageCollections/Storage/PackageCollectionsProfileStorage.swift

```diff
@@ -0,0 +1,339 @@
+                in profile: PackageCollectionsModel.Profile?,
+                callback: @escaping (Result<Bool, Error>) -> Void)
+}
+
+// MARK: - FilePackageCollectionsProfileStorage
+
+struct FilePackageCollectionsProfileStorage: PackageCollectionsProfileStorage {
+    let fileSystem: FileSystem
+    let path: AbsolutePath
+
+    private let jsonEncoder: JSONEncoder
+    private let jsonDecoder: JSONDecoder
+
+    private let lock: LockProtocol
+    private let queue = DispatchQueue(label: "org.swift.swiftpm.FilePackageCollectionsProfileStorage")
+
+    init(fileSystem: FileSystem = localFileSystem, path: AbsolutePath? = nil) {
+        self.fileSystem = fileSystem
+
+        let name = "collections"
+        self.path = path ?? Paths.dotSwiftPM(fileSystem: fileSystem).appending(component: "\(name).json")
+
+        // FileLock only works with real FS
+        if fileSystem.isLocalFileSystem {
+            self.lock = FileLock(name: name, cachePath: self.path.parentDirectory)
+        } else {
+            self.lock = Lock()
+        }
+
+        self.jsonEncoder = JSONEncoder()
+        self.jsonEncoder.outputFormatting = .prettyPrinted
+        if #available(macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0, *) {
+            self.jsonEncoder.outputFormatting.insert(.sortedKeys)
+            self.jsonEncoder.outputFormatting.insert(.withoutEscapingSlashes)
+        }
+        self.jsonDecoder = JSONDecoder()
+    }
+
+    func listProfiles(callback: @escaping (Result<[PackageCollectionsModel.Profile], Error>) -> Void) {
+        self.queue.async {
+            do {
+                let profiles = try self.lock.withLock {
+                    try self.loadFromDisk()
+                }
+                callback(.success(Array(profiles.keys)))
+            } catch {
+                callback(.failure(error))
+            }
+        }
+    }
+
+    func listSources(in profile: PackageCollectionsModel.Profile, callback: @escaping (Result<[PackageCollectionsModel.PackageCollectionSource], Error>) -> Void) {
+        self.queue.async {
+            do {
+                let profiles = try self.lock.withLock {
+                    try self.loadFromDisk()
+                }
+                callback(.success(profiles[profile] ?? []))
+            } catch {
+                callback(.failure(error))
+            }
+        }
+    }
+
+    func add(source: PackageCollectionsModel.PackageCollectionSource,
+             order: Int?,
+             to profile: PackageCollectionsModel.Profile,
+             callback: @escaping (Result<Void, Error>) -> Void) {
+        self.queue.async {
+            do {
+                try self.lock.withLock {
+                    var profiles = try self.loadFromDisk()
+                    var sources = profiles[profile]?.filter { $0 != source } ?? []
+                    let order = order.flatMap { $0 >= 0 && $0 < sources.endIndex ? order : sources.endIndex } ?? sources.endIndex
```

Comment:
Should this return an Error if the index is out-of-bounds, similar to what subscript accessors do?  Or should this be more lenient?

Reply:
@abertelrud I think lenient is fine